File: implementations/python/agent.py
import json
import os
import time
from dataclasses import dataclass, field
from typing import List, Dict, Any, Optional

from openai import OpenAI
from registry import registry

SPEC_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "..", "spec"))


@dataclass
class TurnResult:
    """Structured result returned after completing an agent turn."""
    final_response: str
    messages: List[Dict[str, Any]]
    api_calls: int = 0
    completed: bool = False
    failed: bool = False
    interrupted: bool = False
    exit_reason: str = "unknown"
    error: Optional[str] = None


class Agent:
    def __init__(
        self,
        api_key: Optional[str] = None,
        model: str = "gpt-4-turbo",
        max_iterations: int = 10,
        max_retries: int = 3,
        context_window_limit: int = 30,
    ):
        self.client = OpenAI(api_key=api_key)
        self.model = model
        self.max_iterations = max_iterations
        self.max_retries = max_retries
        self.context_window_limit = context_window_limit
        self.messages: List[Dict[str, Any]] = []
        self._interrupt_requested: bool = False
        
        self._initialize_system_prompt()

    def _initialize_system_prompt(self):
        system_prompt = (
            "You are a helpful AI assistant. You have access to various tools. "
            "When asked to perform a task, use the tools to gather information "
            "and take actions before answering."
        )
        self.messages.append({"role": "system", "content": system_prompt})

    def request_interrupt(self):
        """Signal the agent turn to interrupt on the next iteration check."""
        self._interrupt_requested = True

    def _prepare_api_messages(self, messages: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        """Phase 2.2: Prepare API-only shallow copy of messages with role/sequence validation."""
        api_messages = []
        for msg in messages:
            api_msg = msg.copy()
            # Strip internal metadata fields if present
            api_msg.pop("_internal_id", None)
            api_messages.append(api_msg)
        return api_messages

    def _compress_context_if_needed(self, messages: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        """Phase 2.3: Context window protection.
        
        Trims middle history if messages exceed context_window_limit while preserving:
        - Index 0: System Prompt
        - Index 1: Initial User Prompt (if present)
        - Last N: Recent conversation history
        """
        if len(messages) <= self.context_window_limit:
            return messages

        print(f"  [Context Window Protection] History size ({len(messages)}) > limit ({self.context_window_limit}). Trimming middle history...")
        
        system_prompt = messages[0]
        recent_count = self.context_window_limit - 3
        recent_messages = messages[-recent_count:]

        # Ensure recent_messages starts with a valid role (not an orphaned tool result)
        while recent_messages and recent_messages[0].get("role") == "tool":
            recent_messages.pop(0)

        summary_msg = {
            "role": "system",
            "content": f"[System: Previous conversation history was trimmed to fit context window. {len(messages) - len(recent_messages) - 1} earlier messages summarized.]",
        }
        
        return [system_prompt, summary_msg] + recent_messages

    def run(self, user_input: str) -> TurnResult:
        """Phase 1 -> Phase 2 -> Phase 3 -> Phase 4 Conversation Loop execution."""
        # --- Phase 1: Turn Prologue ---
        print(f"\nUser: {user_input}")
        self.messages.append({"role": "user", "content": user_input})
        
        api_call_count = 0
        self._interrupt_requested = False
        empty_content_retries = 0

        # --- Phase 2: Main Conversation Loop ---
        while api_call_count < self.max_iterations:
            # 2.1 Pre-API Checks
            if self._interrupt_requested:
                print("  [Turn Exit] Turn interrupted by user.")
                return TurnResult(
                    final_response="",
                    messages=self.messages,
                    api_calls=api_call_count,
                    interrupted=True,
                    exit_reason="interrupted",
                )

            api_call_count += 1

            # 2.2 & 2.3 Message Preparation and Context Compression
            prepared_messages = self._prepare_api_messages(self.messages)
            prepared_messages = self._compress_context_if_needed(prepared_messages)

            # 2.4 Inner Retry Loop for LLM API Call
            response = None
            for retry in range(self.max_retries):
                try:
                    tool_schemas = registry.get_tool_schemas()
                    response = self.client.chat.completions.create(
                        model=self.model,
                        messages=prepared_messages,
                        tools=tool_schemas if tool_schemas else None,
                    )
                    break
                except Exception as e:
                    print(f"  [API Error Retry {retry + 1}/{self.max_retries}] {e}")
                    if retry == self.max_retries - 1:
                        return TurnResult(
                            final_response="",
                            messages=self.messages,
                            api_calls=api_call_count,
                            failed=True,
                            exit_reason="api_error",
                            error=str(e),
                        )
                    time.sleep(2 ** retry)

            message = response.choices[0].message
            message_dict = message.model_dump(exclude_unset=True)
            
            # Append assistant message to canonical history
            self.messages.append(message_dict)

            # 2.6 Tool Call Execution Path
            if message.tool_calls:
                for tool_call in message.tool_calls:
                    name = tool_call.function.name
                    call_id = tool_call.id
                    
                    # Validate tool exists in registry (Self-correction path)
                    available_tools = list(registry._tools.keys())
                    if name not in registry._tools:
                        error_msg = f"Error: Tool '{name}' is not registered. Available tools: {available_tools}"
                        print(f"  [Tool Validation Error] {error_msg}")
                        self.messages.append({
                            "role": "tool",
                            "tool_call_id": call_id,
                            "name": name,
                            "content": error_msg,
                        })
                        continue

                    # Validate JSON arguments
                    try:
                        args = json.loads(tool_call.function.arguments) if tool_call.function.arguments else {}
                    except json.JSONDecodeError as json_err:
                        error_msg = f"Error: Invalid JSON arguments for tool '{name}': {json_err}"
                        print(f"  [JSON Parse Error] {error_msg}")
                        self.messages.append({
                            "role": "tool",
                            "tool_call_id": call_id,
                            "name": name,
                            "content": error_msg,
                        })
                        continue

                    print(f"  [Tool Execution] {name}({args})")
                    
                    # Execute tool with exception handling
                    try:
                        result = registry.execute_tool(name, args)
                        print(f"  [Tool Result] {result}")
                    except Exception as exec_err:
                        result = f"Error executing tool '{name}': {exec_err}"
                        print(f"  [Tool Runtime Error] {result}")
                    
                    self.messages.append({
                        "role": "tool",
                        "tool_call_id": call_id,
                        "name": name,
                        "content": str(result),
                    })
                
                # Continue loop to send tool results back to LLM
                continue

            # 2.7 Final Text Response Path
            final_text = (message.content or "").strip()
            
            # Empty Response Recovery
            if not final_text:
                if empty_content_retries < 2:
                    empty_content_retries += 1
                    print("  [Empty Response Recovery] Retrying with prompt nudge...")
                    self.messages.append({
                        "role": "user",
                        "content": "Please provide a complete text response summarizing your answer.",
                    })
                    continue
                else:
                    final_text = "(empty response)"

            print(f"Assistant: {final_text}")
            
            # --- Phase 4: Turn Finalization ---
            return TurnResult(
                final_response=final_text,
                messages=self.messages,
                api_calls=api_call_count,
                completed=True,
                exit_reason="text_response",
            )

        # Exceeded iteration budget
        print(f"  [Turn Exit] Reached max iterations ({self.max_iterations}).")
        return TurnResult(
            final_response="Reached maximum iteration limit.",
            messages=self.messages,
            api_calls=api_call_count,
            completed=False,
            failed=True,
            exit_reason="budget_exhausted",
        )

File: implementations/typescript/src/Agent.ts
import OpenAI from 'openai';
import { registry } from './ToolRegistry';

export interface TurnResult {
  finalResponse: string;
  messages: OpenAI.Chat.Completions.ChatCompletionMessageParam[];
  apiCalls: number;
  completed: boolean;
  failed: boolean;
  interrupted: boolean;
  exitReason: string;
  error?: string;
}

export interface AgentConfig {
  apiKey?: string;
  model?: string;
  maxIterations?: number;
  maxRetries?: number;
  contextWindowLimit?: number;
}

export class Agent {
  private client: OpenAI;
  private model: string;
  private maxIterations: number;
  private maxRetries: number;
  private contextWindowLimit: number;
  private messages: OpenAI.Chat.Completions.ChatCompletionMessageParam[] = [];
  private interruptRequested: boolean = false;

  constructor(apiKey?: string, model: string = 'gpt-4-turbo', config: AgentConfig = {}) {
    this.client = new OpenAI({ apiKey: config.apiKey || apiKey });
    this.model = config.model || model;
    this.maxIterations = config.maxIterations ?? 10;
    this.maxRetries = config.maxRetries ?? 3;
    this.contextWindowLimit = config.contextWindowLimit ?? 30;
    this.initializeSystemPrompt();
  }

  private initializeSystemPrompt() {
    const systemPrompt = `You are a helpful AI assistant. You have access to various tools. 
When asked to perform a task, use the tools to gather information and take actions before answering.`;
    this.messages.push({ role: 'system', content: systemPrompt });
  }

  public requestInterrupt(): void {
    this.interruptRequested = true;
  }

  public prepareApiMessages(
    msgs: OpenAI.Chat.Completions.ChatCompletionMessageParam[]
  ): OpenAI.Chat.Completions.ChatCompletionMessageParam[] {
    // Phase 2.2: Shallow copy for API payload to keep canonical history clean
    return msgs.map((m) => ({ ...m }));
  }

  public compressContextIfNeeded(
    msgs: OpenAI.Chat.Completions.ChatCompletionMessageParam[]
  ): OpenAI.Chat.Completions.ChatCompletionMessageParam[] {
    // Phase 2.3: Context window protection
    if (msgs.length <= this.contextWindowLimit) {
      return msgs;
    }

    console.log(
      `  [Context Window Protection] History size (${msgs.length}) > limit (${this.contextWindowLimit}). Trimming middle history...`
    );

    const systemPrompt = msgs[0];
    const recentCount = this.contextWindowLimit - 3;
    const recentMessages = msgs.slice(-recentCount);

    while (recentMessages.length > 0 && recentMessages[0].role === 'tool') {
      recentMessages.shift();
    }

    const summaryMsg: OpenAI.Chat.Completions.ChatCompletionMessageParam = {
      role: 'system',
      content: `[System: Previous conversation history was trimmed to fit context window. ${
        msgs.length - recentMessages.length - 1
      } earlier messages summarized.]`,
    };

    return [systemPrompt, summaryMsg, ...recentMessages];
  }

  async run(userInput: string): Promise<TurnResult> {
    // --- Phase 1: Turn Prologue ---
    console.log(`\nUser: ${userInput}`);
    this.messages.push({ role: 'user', content: userInput });

    let apiCalls = 0;
    let emptyContentRetries = 0;

    // --- Phase 2: Main Conversation Loop ---
    while (apiCalls < this.maxIterations) {
      // 2.1 Pre-API Checks
      if (this.interruptRequested) {
        this.interruptRequested = false;
        console.log('  [Turn Exit] Turn interrupted by user.');
        return {
          finalResponse: '',
          messages: this.messages,
          apiCalls,
          completed: false,
          failed: false,
          interrupted: true,
          exitReason: 'interrupted',
        };
      }

      apiCalls++;

      // 2.2 & 2.3 Message Preparation and Context Compression
      let preparedMessages = this.prepareApiMessages(this.messages);
      preparedMessages = this.compressContextIfNeeded(preparedMessages);

      // 2.4 Inner Retry Loop for LLM API Call
      let response: OpenAI.Chat.Completions.ChatCompletion | null = null;
      let lastError: any = null;

      for (let retry = 0; retry < this.maxRetries; retry++) {
        try {
          const tools = registry.getToolSchemas();
          response = await this.client.chat.completions.create({
            model: this.model,
            messages: preparedMessages,
            tools: tools.length > 0 ? tools : undefined,
          });
          break;
        } catch (err: any) {
          lastError = err;
          console.log(`  [API Error Retry ${retry + 1}/${this.maxRetries}] ${err.message || err}`);
          if (retry === this.maxRetries - 1) {
            return {
              finalResponse: '',
              messages: this.messages,
              apiCalls,
              completed: false,
              failed: true,
              interrupted: false,
              exitReason: 'api_error',
              error: String(err.message || err),
            };
          }
          await new Promise((resolve) => setTimeout(resolve, Math.pow(2, retry) * 1000));
        }
      }

      if (!response) {
        return {
          finalResponse: '',
          messages: this.messages,
          apiCalls,
          completed: false,
          failed: true,
          interrupted: false,
          exitReason: 'no_response',
          error: String(lastError),
        };
      }

      const message = response.choices[0].message;
      this.messages.push(message);

      // 2.6 Tool Call Execution Path
      if (message.tool_calls && message.tool_calls.length > 0) {
        for (const toolCall of message.tool_calls) {
          const name = toolCall.function.name;
          const callId = toolCall.id;

          // Self-correction for unknown tools
          const registeredTools = registry.getToolSchemas().map((t) => t.function.name);
          if (!registeredTools.includes(name)) {
            const errorMsg = `Error: Tool '${name}' is not registered. Available tools: ${registeredTools.join(', ')}`;
            console.log(`  [Tool Validation Error] ${errorMsg}`);
            this.messages.push({
              role: 'tool',
              tool_call_id: callId,
              content: errorMsg,
            });
            continue;
          }

          // Parse JSON arguments
          let args: Record<string, any> = {};
          try {
            args = toolCall.function.arguments ? JSON.parse(toolCall.function.arguments) : {};
          } catch (jsonErr: any) {
            const errorMsg = `Error: Invalid JSON arguments for tool '${name}': ${jsonErr.message}`;
            console.log(`  [JSON Parse Error] ${errorMsg}`);
            this.messages.push({
              role: 'tool',
              tool_call_id: callId,
              content: errorMsg,
            });
            continue;
          }

          console.log(`  [Tool Execution] ${name}(${JSON.stringify(args)})`);

          // Execute tool with runtime exception handling
          try {
            const result = await registry.executeTool(name, args);
            console.log(`  [Tool Result] ${result}`);
            this.messages.push({
              role: 'tool',
              tool_call_id: callId,
              content: String(result),
            });
          } catch (execErr: any) {
            const result = `Error executing tool '${name}': ${execErr.message || execErr}`;
            console.log(`  [Tool Runtime Error] ${result}`);
            this.messages.push({
              role: 'tool',
              tool_call_id: callId,
              content: result,
            });
          }
        }
        // Continue loop to send tool results to LLM
        continue;
      }

      // 2.7 Final Text Response Path
      let finalText = (message.content || '').trim();

      // Empty Response Recovery
      if (!finalText) {
        if (emptyContentRetries < 2) {
          emptyContentRetries++;
          console.log('  [Empty Response Recovery] Retrying with prompt nudge...');
          this.messages.push({
            role: 'user',
            content: 'Please provide a complete text response summarizing your answer.',
          });
          continue;
        } else {
          finalText = '(empty response)';
        }
      }

      console.log(`Assistant: ${finalText}`);

      // --- Phase 4: Turn Finalization ---
      return {
        finalResponse: finalText,
        messages: this.messages,
        apiCalls,
        completed: true,
        failed: false,
        interrupted: false,
        exitReason: 'text_response',
      };
    }

    // Exceeded iteration budget
    console.log(`  [Turn Exit] Reached max iterations (${this.maxIterations}).`);
    return {
      finalResponse: 'Reached maximum iteration limit.',
      messages: this.messages,
      apiCalls,
      completed: false,
      failed: true,
      interrupted: false,
      exitReason: 'budget_exhausted',
    };
  }
}
